File: monitorS/client_apps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
软件使用监控模块
- 统计各软件使用时长
- 前台应用识别
"""

import time
import logging
import threading
from datetime import datetime
from typing import Dict, List, Optional
import json
from pathlib import Path

try:
    import psutil

    PSUTIL_AVAILABLE = True
except ImportError:
    PSUTIL_AVAILABLE = False

try:
    import win32gui
    import win32process

    WIN32_AVAILABLE = True
except ImportError:
    WIN32_AVAILABLE = False


class AppMonitor:
    """软件使用监控器"""

    # 忽略的系统进程
    IGNORE_PROCESSES = [
        # Windows 核心进程
        "System",
        "Registry",
        "smss.exe",
        "csrss.exe",
        "wininit.exe",
        "services.exe",
        "lsass.exe",
        "lsm.exe",
        "svchost.exe",
        "conhost.exe",
        "fontdrvhost.exe",
        "dwm.exe",
        "Idle",
        # 后台服务进程
        "dllhost.exe",
        "RuntimeBroker.exe",
        "backgroundTaskHost.exe",
        "sihost.exe",
        "taskhostw.exe",
        "ShellExperienceHost.exe",
        "SearchUI.exe",
        "SearchIndexer.exe",
        "WmiPrvSE.exe",
        "spoolsv.exe",
        # 数据库服务
        "postgres.exe",
        "mysqld.exe",
        "sqlservr.exe",
        "mongod.exe",
        "redis-server.exe",
        # 开发工具后台
        "node.exe",
        "python.exe",
        "java.exe",
        "javaw.exe",
        "git.exe",
        # 系统工具
        "cmd.exe",
        "powershell.exe",
        "powershell_ise.exe",
        "explorer.exe",
        "taskmgr.exe",
        "regedit.exe",
        "msconfig.exe",
        # 驱动程序
        "audiodg.exe",
        "rundll32.exe",
        "CompatTelRunner.exe",
        # 更新程序
        "wuauclt.exe",
        "TrustedInstaller.exe",
        "TiWorker.exe",
        # 杀毒软件
        "MsMpEng.exe",
        "NisSrv.exe",
        "SecurityHealthService.exe",
        "SecurityHealthSystray.exe",
    ]

    def __init__(self, client):
        self.client = client
        self.logger = logging.getLogger(__name__)

        # 活跃应用 {pid: {"name": name, "start": time, "last_seen": time}}
        self.active_apps = {}
        self.lock = threading.RLock()

        # 上报间隔
        self.report_interval = 300  # 5分钟
        self.last_report_time = time.time()

        # 缓存
        self.stats_cache = []

    def get_all_processes(self) -> List[Dict]:
        """获取所有进程"""
        processes = []
        if not PSUTIL_AVAILABLE:
            return processes

        for proc in psutil.process_iter(
            ["pid", "name", "exe", "create_time", "cpu_percent", "memory_info"]
        ):
            try:
                pinfo = proc.info
                if pinfo["name"] and pinfo["name"] not in self.IGNORE_PROCESSES:
                    processes.append(
                        {
                            "pid": pinfo["pid"],
                            "name": pinfo["name"],
                            "exe": pinfo["exe"] or "",
                            "create_time": pinfo["create_time"],
                            "cpu_percent": proc.cpu_percent(),
                            "memory_mb": (
                                pinfo["memory_info"].rss / 1024 / 1024
                                if pinfo["memory_info"]
                                else 0
                            ),
                        }
                    )
            except (psutil.NoSuchProcess, psutil.AccessDenied, AttributeError):
                continue

        return processes

    def get_foreground_app(self) -> Optional[Dict]:
        """获取当前前台应用"""
        if not WIN32_AVAILABLE or not PSUTIL_AVAILABLE:
            return None

        try:
            hwnd = win32gui.GetForegroundWindow()
            _, pid = win32process.GetWindowThreadProcessId(hwnd)

            proc = psutil.Process(pid)
            proc_name = proc.name()

            if proc_name not in self.IGNORE_PROCESSES:
                return {
                    "pid": pid,
                    "name": proc_name,
                    "exe": proc.exe(),
                    "window_title": win32gui.GetWindowText(hwnd),
                    "cpu_percent": proc.cpu_percent(),
                    "memory_mb": proc.memory_info().rss / 1024 / 1024,
                    "timestamp": time.time(),
                }
        except Exception as e:
            self.logger.debug(f"获取前台应用失败: {e}")

        return None

    def track_app_usage(self):
        """跟踪软件使用 - 智能识别（只记录真实用户软件）"""
        current_time = time.time()
        foreground = self.get_foreground_app()

        # 将忽略列表转为小写，便于比较
        ignore_lower = [p.lower() for p in self.IGNORE_PROCESSES]

        with self.lock:
            # ===== 1. 先记录所有应用（但过滤系统进程）=====
            for proc in self.get_all_processes():
                pid = proc["pid"]
                proc_name = proc["name"].lower()

                # 过滤系统进程
                if proc_name in ignore_lower:
                    continue

                # 忽略空名称
                if not proc_name:
                    continue

                if pid in self.active_apps:
                    # 更新最后活跃时间
                    self.active_apps[pid]["last_seen"] = current_time
                    # 更新资源使用（每秒采样）
                    self.active_apps[pid]["cpu_samples"] = self.active_apps[pid].get(
                        "cpu_samples", []
                    ) + [proc["cpu_percent"]]
                    self.active_apps[pid]["memory_samples"] = self.active_apps[pid].get(
                        "memory_samples", []
                    ) + [proc["memory_mb"]]
                else:
                    # 新应用
                    self.active_apps[pid] = {
                        "pid": pid,
                        "name": proc["name"],
                        "exe": proc["exe"],
                        "first_seen": current_time,
                        "last_seen": current_time,
                        "foreground_time": 0,
                        "total_time": 0,
                        "cpu_samples": [proc["cpu_percent"]],
                        "memory_samples": [proc["memory_mb"]],
                    }

            # ===== 2. 标记前台应用（增加权重）=====
            if foreground and foreground["pid"]:
                pid = foreground["pid"]
                app_name = foreground["name"].lower()

                # 如果前台应用是系统进程，忽略
                if app_name in ignore_lower:
                    pass  # 不记录系统进程
                elif pid in self.active_apps:
                    # 已有记录，增加前台时间
                    self.active_apps[pid]["foreground_time"] += 5
                    self.active_apps[pid]["last_seen"] = current_time
                    # 更新窗口标题（前台应用才有）
                    self.active_apps[pid]["window_title"] = foreground.get(
                        "window_title", ""
                    )
                    self.active_apps[pid]["cpu_percent"] = foreground.get(
                        "cpu_percent", 0
                    )
                    self.active_apps[pid]["memory_mb"] = foreground.get("memory_mb", 0)
                else:
                    # 新前台应用（虽然被过滤了，但还是记录一下）
                    self.active_apps[pid] = {
                        "pid": pid,
                        "name": foreground["name"],
                        "exe": foreground["exe"],
                        "first_seen": current_time,
                        "last_seen": current_time,
                        "foreground_time": 5,
                        "total_time": 0,
                        "window_title": foreground.get("window_title", ""),
                        "cpu_percent": foreground.get("cpu_percent", 0),
                        "memory_mb": foreground.get("memory_mb", 0),
                        "cpu_samples": [foreground.get("cpu_percent", 0)],
                        "memory_samples": [foreground.get("memory_mb", 0)],
                    }

            # ===== 3. 清理不活跃应用（30秒无活动）=====
            to_remove = []
            for pid, info in self.active_apps.items():
                if current_time - info["last_seen"] > 30:
                    # 计算使用时长
                    total_duration = int(current_time - info["first_seen"])

                    # 智能判断：满足以下任一条件才记录
                    should_save = False

                    # 条件1：运行时间超过30秒
                    if total_duration >= 30:
                        should_save = True

                    # 条件2：前台时间超过10秒
                    if info.get("foreground_time", 0) >= 10:
                        should_save = True

                    # 条件3：是浏览器或办公软件（通过名称判断）
                    user_apps = [
                        "chrome",
                        "firefox",
                        "edge",
                        "brave",
                        "vivaldi",
                        "navicat",
                        "notepad",
                        "code",
                        "word",
                        "excel",
                        "powerpnt",
                        "wechat",
                        "qq",
                        "dingtalk",
                    ]
                    if any(app in info["name"].lower() for app in user_apps):
                        should_save = True

                    if should_save:
                        self._save_app_usage(info, total_duration)

                    to_remove.append(pid)

            # 删除已处理的进程
            for pid in to_remove:
                del self.active_apps[pid]

    def _save_app_usage(self, info: Dict, duration: int):
        """保存软件使用记录 - 只记录前台应用"""
        if not self.client.employee_id:
            return

        # ✅ 只记录前台应用（前台时间占比>30%）
        foreground_ratio = info.get("foreground_time", 0) / max(duration, 1)
        if foreground_ratio < 0.3:
            self.logger.debug(
                f"跳过后台应用: {info['name']} (前台占比{foreground_ratio:.1%})"
            )
            return

        # 计算平均值
        cpu_samples = info.get("cpu_samples", [0])
        memory_samples = info.get("memory_samples", [0])

        cpu_avg = int(sum(cpu_samples) / len(cpu_samples) if cpu_samples else 0)
        memory_avg = int(
            sum(memory_samples) / len(memory_samples) if memory_samples else 0
        )

        data = {
            "employee_id": self.client.employee_id,
            "client_id": self.client.client_id,
            "app_name": info["name"],
            "app_path": info.get("exe", ""),
            "window_title": info.get("window_title", ""),
            "start_time": datetime.fromtimestamp(info["first_seen"]),
            "end_time": datetime.fromtimestamp(info["last_seen"]),
            "duration": duration,
            "is_foreground": True,  # ✅ 始终为 True
            "cpu_avg": cpu_avg,
            "memory_avg": memory_avg,
        }

        # 添加到缓存
        self.stats_cache.append(data)

        # 保存到文件（备份）
        cache_file = Path("cache") / f"app_{info['first_seen']}.json"
        cache_file.parent.mkdir(exist_ok=True)

        data_for_file = data.copy()
        data_for_file["start_time"] = data["start_time"].isoformat()
        data_for_file["end_time"] = data["end_time"].isoformat()

        with open(cache_file, "w", encoding="utf-8") as f:
            json.dump(data_for_file, f)

        # 调试日志
        self.logger.debug(
            f"📝 记录前台应用: {info['name']} - {duration}秒 (前台占比{foreground_ratio:.1%})"
        )

    def collect_and_report(self):
        """收集并上报软件使用统计"""
        current_time = time.time()

        if current_time - self.last_report_time < self.report_interval:
            return

        if not self.stats_cache:
            return

        foreground_count = sum(
            1 for item in self.stats_cache if item.get("is_foreground")
        )
        background_count = len(self.stats_cache) - foreground_count
        self.logger.info(
            f"开始上报软件使用统计: 共{len(self.stats_cache)}条 "
            f"(前台:{foreground_count}, 后台:{background_count})"
        )

        top_apps = [item["app_name"] for item in self.stats_cache[:5]]
        self.logger.info(f"应用示例: {', '.join(top_apps)}")

        # ✅ 关键修复：确保所有 datetime 对象都转为字符串
        serializable_cache = []
        for item in self.stats_cache:
            serializable_item = item.copy()
            if isinstance(serializable_item.get("start_time"), datetime):
                serializable_item["start_time"] = serializable_item[
                    "start_time"
                ].isoformat()
            if isinstance(serializable_item.get("end_time"), datetime):
                serializable_item["end_time"] = serializable_item[
                    "end_time"
                ].isoformat()
            serializable_cache.append(serializable_item)

        if self.client.api_client and not self.client.offline_mode:
            try:
                response = self.client.api_client.post(
                    "/api/apps/usage", json=serializable_cache  # ✅ 使用转换后的数据
                )
                if response:
                    self.logger.info(
                        f"✅ 软件使用上报成功: {len(serializable_cache)}条"
                    )
                    # 删除已上报的缓存文件
                    for item in self.stats_cache:
                        # 确保从对象获取时间戳用于定位文件名
                        if isinstance(item["start_time"], datetime):
                            start = item["start_time"].timestamp()
                        else:
                            start = datetime.fromisoformat(
                                item["start_time"]
                            ).timestamp()

                        cache_file = Path("cache") / f"app_{start}.json"
                        if cache_file.exists():
                            cache_file.unlink()
                    self.stats_cache.clear()
                    self.last_report_time = current_time
                else:
                    self.logger.warning("上报失败，保留缓存")
            except Exception as e:
                self.logger.error(f"上报软件使用失败: {e}")
        else:
            self.logger.debug("离线模式，软件使用数据已缓存")

    def start_monitoring(self):
        """启动监控线程"""

        def monitor_loop():
            while self.client.running:
                try:
                    self.track_app_usage()
                    self.collect_and_report()
                    time.sleep(5)
                except Exception as e:
                    self.logger.error(f"软件监控异常: {e}")
                    time.sleep(30)

        thread = threading.Thread(target=monitor_loop, daemon=True)
        thread.start()
        self.logger.info("✅ 软件使用监控已启动")

File: monitor-go/client_apps.go
// client_apps.go - 软件使用监控器（实时采集版）
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/shirou/gopsutil/v3/process"
)

// AppMonitor 软件使用监控器
type AppMonitor struct {
	client          *MonitorClient
	ignoreProcesses []string
	activeApps      map[int]*ActiveApp
	reportInterval  time.Duration
	lastReportTime  time.Time
	running         int32
	mu              sync.RWMutex
	stopCh          chan struct{}
	wg              sync.WaitGroup

	// 配置
	maxCacheSize int
}

// ActiveApp 活跃应用
type ActiveApp struct {
	PID            int
	Name           string
	Exe            string
	StartTime      time.Time
	LastSeen       time.Time
	WindowTitle    string
	CPUPercent     float64
	MemoryMB       float64
}

// AppUsage 软件使用记录
type AppUsage struct {
	EmployeeID  string    `json:"employee_id"`
	ClientID    string    `json:"client_id"`
	AppName     string    `json:"app_name"`
	AppPath     string    `json:"app_path"`
	WindowTitle string    `json:"window_title"`
	StartTime   time.Time `json:"start_time"`
	Duration    int       `json:"duration"`
	CPUAvg      int       `json:"cpu_avg"`
	MemoryAvg   int       `json:"memory_avg"`
}

// NewAppMonitor 创建软件监控器
func NewAppMonitor(client *MonitorClient) *AppMonitor {
	return &AppMonitor{
		client: client,
		ignoreProcesses: []string{
			"System", "Registry", "smss.exe", "csrss.exe", "wininit.exe",
			"services.exe", "lsass.exe", "lsm.exe", "svchost.exe", "conhost.exe",
			"fontdrvhost.exe", "dwm.exe", "Idle", "dllhost.exe", "RuntimeBroker.exe",
			"backgroundTaskHost.exe", "sihost.exe", "taskhostw.exe", "ShellExperienceHost.exe",
			"SearchUI.exe", "SearchIndexer.exe", "WmiPrvSE.exe", "spoolsv.exe",
			"postgres.exe", "mysqld.exe", "sqlservr.exe", "mongod.exe", "redis-server.exe",
			"node.exe", "python.exe", "java.exe", "javaw.exe", "git.exe",
			"cmd.exe", "powershell.exe", "explorer.exe", "taskmgr.exe",
		},
		activeApps:     make(map[int]*ActiveApp),
		reportInterval: 120 * time.Second, // 2分钟上报一次
		stopCh:         make(chan struct{}),
		maxCacheSize:   10000,
	}
}

// StartMonitoring 启动监控
func (a *AppMonitor) StartMonitoring() {
	if !atomic.CompareAndSwapInt32(&a.running, 0, 1) {
		logInfo("软件使用监控器已在运行")
		return
	}

	logInfo("软件使用监控器启动 [上报间隔: %v, 实时采集模式]", a.reportInterval)
	a.wg.Add(1)
	go a.monitorLoop()
}

// StopMonitoring 停止监控
func (a *AppMonitor) StopMonitoring() {
	if !atomic.CompareAndSwapInt32(&a.running, 1, 0) {
		return
	}

	logInfo("软件使用监控器停止中...")
	close(a.stopCh)

	// 保存所有未保存的应用
	a.flushActiveApps()

	done := make(chan struct{})
	go func() {
		a.wg.Wait()
		close(done)
	}()

	select {
	case <-done:
		logInfo("软件使用监控器已停止")
	case <-time.After(5 * time.Second):
		logWarn("软件使用监控器停止超时")
	}
}

// flushActiveApps 刷新所有活跃应用（停止时调用）
func (a *AppMonitor) flushActiveApps() {
	a.mu.Lock()
	defer a.mu.Unlock()

	for _, info := range a.activeApps {
		duration := int(info.LastSeen.Sub(info.StartTime).Seconds())
		if duration >= 5 {
			logInfo("⏹️ 停止时保存应用: %s (时长: %d秒)", info.Name, duration)
			a.saveAppUsage(info, duration)
		}
	}
	a.activeApps = make(map[int]*ActiveApp)
}

func (a *AppMonitor) monitorLoop() {
	defer a.wg.Done()

	logInfo("软件使用监控循环已启动")
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			if atomic.LoadInt32(&a.client.running) == 0 {
				return
			}
			a.trackAppUsage()
			a.collectAndReport()
		case <-a.stopCh:
			return
		}
	}
}

// trackAppUsage 跟踪应用使用（实时采集）
func (a *AppMonitor) trackAppUsage() {
	currentTime := time.Now()
	foreground := a.getForegroundApp()

	a.mu.Lock()
	defer a.mu.Unlock()

	// 检查前台窗口是否是应用（排除忽略列表）
	isAppForeground := foreground != nil && foreground.PID > 0

	if isAppForeground {
		pid := foreground.PID
		appName := foreground.Name

		if existing, ok := a.activeApps[pid]; ok {
			// 更新最后活跃时间
			existing.LastSeen = currentTime
			existing.CPUPercent = foreground.CPUPercent
			existing.MemoryMB = foreground.MemoryMB
			existing.WindowTitle = foreground.WindowTitle

			// 检查应用是否变化（同一个PID可能运行不同程序？这种情况很少）
			// 主要是更新窗口标题
		} else {
			// 新应用，创建记录
			a.activeApps[pid] = &ActiveApp{
				PID:         pid,
				Name:        appName,
				Exe:         foreground.Exe,
				StartTime:   currentTime,
				LastSeen:    currentTime,
				WindowTitle: foreground.WindowTitle,
				CPUPercent:  foreground.CPUPercent,
				MemoryMB:    foreground.MemoryMB,
			}
			logDebug("🆕 开始跟踪应用: %s - %s", appName, foreground.WindowTitle)
		}
	} else {
		// 当前不是应用窗口（桌面、任务管理器等），保存所有正在跟踪的应用
		if len(a.activeApps) > 0 {
			logDebug("📴 用户离开应用，保存所有使用记录")
			for _, info := range a.activeApps {
				duration := int(info.LastSeen.Sub(info.StartTime).Seconds())
				if duration >= 5 {
					logDebug("⏹️ 保存应用: %s - %s (时长: %d秒)", info.Name, info.WindowTitle, duration)
					a.saveAppUsage(info, duration)
				} else if duration < 5 {
					logDebug("丢弃短时间使用: %s (时长: %d秒)", info.Name, duration)
				}
			}
			// 清空所有活跃记录
			a.activeApps = make(map[int]*ActiveApp)
		}
		return
	}

	// 清理超时记录（超过30秒未更新，作为兜底）
	toRemove := make([]int, 0)
	for pid, info := range a.activeApps {
		if currentTime.Sub(info.LastSeen) > 30*time.Second {
			duration := int(info.LastSeen.Sub(info.StartTime).Seconds())
			if duration >= 5 {
				logDebug("⏹️ 超时保存应用: %s - %s (时长: %d秒)", info.Name, info.WindowTitle, duration)
				a.saveAppUsage(info, duration)
			}
			toRemove = append(toRemove, pid)
		}
	}

	for _, pid := range toRemove {
		delete(a.activeApps, pid)
	}
}

// getAllProcesses 获取所有进程（用于更新CPU/内存样本）
func (a *AppMonitor) getAllProcesses() []ProcessInfo {
	if runtime.GOOS != "windows" {
		return []ProcessInfo{}
	}

	ps, err := process.Processes()
	if err != nil {
		logDebug("获取进程列表失败: %v", err)
		return []ProcessInfo{}
	}

	processes := make([]ProcessInfo, 0, len(ps))
	for _, p := range ps {
		if p == nil {
			continue
		}

		name, err := p.Name()
		if err != nil || name == "" {
			continue
		}

		cpu, _ := p.CPUPercent()

		mem, _ := p.MemoryInfo()
		memMB := float64(0)
		if mem != nil {
			memMB = float64(mem.RSS) / 1024 / 1024
		}

		processes = append(processes, ProcessInfo{
			PID:        int(p.Pid),
			Name:       name,
			Exe:        "",
			CPUPercent: cpu,
			MemoryMB:   memMB,
		})
	}

	return processes
}

// getForegroundApp 获取当前前台应用
func (a *AppMonitor) getForegroundApp() *ForegroundApp {
	if runtime.GOOS != "windows" {
		return nil
	}

	winInfo := GetForegroundWindowInfo()
	if winInfo == nil {
		return nil
	}

	p, err := process.NewProcess(int32(winInfo.PID))
	if err != nil {
		return nil
	}

	name, err := p.Name()
	if err != nil || name == "" {
		return nil
	}

	// 检查是否在忽略列表中
	nameLower := strings.ToLower(name)
	for _, ign := range a.ignoreProcesses {
		if nameLower == strings.ToLower(ign) {
			return nil
		}
	}

	cpu, _ := p.CPUPercent()
	mem, _ := p.MemoryInfo()
	memMB := float64(0)
	if mem != nil {
		memMB = float64(mem.RSS) / 1024 / 1024
	}

	return &ForegroundApp{
		PID:         int(winInfo.PID),
		Name:        name,
		Exe:         "",
		WindowTitle: winInfo.Title,
		CPUPercent:  cpu,
		MemoryMB:    memMB,
	}
}

// saveAppUsage 保存应用使用记录
func (a *AppMonitor) saveAppUsage(info *ActiveApp, duration int) {
	if !a.client.IsReady() {
		logDebug("saveAppUsage: 客户端未就绪，跳过保存")
		return
	}

	// 只保存使用超过5秒的应用
	if duration < 5 {
		logDebug("使用时长不足5秒，丢弃: %s (%d秒)", info.Name, duration)
		return
	}

	// 计算CPU和内存平均值（简化：使用最后采样的值）
	cpuAvg := int(info.CPUPercent)
	memoryAvg := int(info.MemoryMB)

	usage := AppUsage{
		EmployeeID:  a.client.employeeID,
		ClientID:    a.client.clientID,
		AppName:     info.Name,
		AppPath:     info.Exe,
		WindowTitle: info.WindowTitle,
		StartTime:   info.StartTime,
		Duration:    duration,
		CPUAvg:      cpuAvg,
		MemoryAvg:   memoryAvg,
	}

	// 保存到缓存文件
	go a.saveUsageToFile(usage)
	logDebug("💾 保存应用记录: [%s] %s (%d秒)", info.Name, info.WindowTitle, duration)
}

// saveUsageToFile 保存使用记录到文件
func (a *AppMonitor) saveUsageToFile(usage AppUsage) {
	cacheDir := "cache"
	if err := os.MkdirAll(cacheDir, 0755); err != nil {
		logError("创建缓存目录失败: %v", err)
		return
	}

	cacheFile := filepath.Join(cacheDir, fmt.Sprintf("app_%d_%d.json", usage.StartTime.Unix(), time.Now().UnixNano()))

	jsonData, err := json.Marshal(usage)
	if err != nil {
		logError("序列化失败: %v", err)
		return
	}

	if err := os.WriteFile(cacheFile, jsonData, 0644); err != nil {
		logError("保存缓存文件失败: %v", err)
	}
}

// collectAndReport 收集并上报软件使用统计
func (a *AppMonitor) collectAndReport() {
	currentTime := time.Now()

	// 检查上报间隔
	if currentTime.Sub(a.lastReportTime) < a.reportInterval {
		return
	}

	if a.client.employeeID == "" {
		logDebug("软件使用上报: employeeID未就绪，跳过")
		return
	}

	logInfo("📊 开始收集软件使用数据...")

	// 收集所有已保存的使用记录
	allUsage := a.readAllSavedUsage()

	if len(allUsage) == 0 {
		logInfo("📊 没有软件使用记录需要上报")
		a.lastReportTime = currentTime
		return
	}

	// 按时间排序（最新的在前）
	sort.Slice(allUsage, func(i, j int) bool {
		return allUsage[i].StartTime.After(allUsage[j].StartTime)
	})

	logInfo("📊 收集到 %d 条软件使用记录", len(allUsage))

	// 打印所有记录
	for i, item := range allUsage {
		logInfo("  %d. [%s] %s (时长: %d秒, 时间: %s)",
			i+1, item.AppName, item.WindowTitle, item.Duration,
			item.StartTime.Format("15:04:05"))
	}

	// 上报所有记录
	if len(allUsage) > 0 {
		a.uploadUsageLocked(allUsage, currentTime)
	} else {
		a.lastReportTime = currentTime
	}
}

// readAllSavedUsage 读取所有已保存的软件使用记录
func (a *AppMonitor) readAllSavedUsage() []AppUsage {
	results := make([]AppUsage, 0)
	cacheDir := "cache"

	// 读取所有软件使用缓存文件
	files, err := filepath.Glob(filepath.Join(cacheDir, "app_*.json"))
	if err != nil {
		return results
	}

	for _, cacheFile := range files {
		data, err := os.ReadFile(cacheFile)
		if err != nil {
			continue
		}

		var item AppUsage
		if err := json.Unmarshal(data, &item); err == nil {
			// 补全ID信息
			item.EmployeeID = a.client.employeeID
			item.ClientID = a.client.clientID

			// 只保留最近2小时内的记录
			if time.Since(item.StartTime) < 2*time.Hour {
				results = append(results, item)
			}
		}
		// 读取后删除缓存文件
		os.Remove(cacheFile)
	}

	return results
}

// uploadUsageLocked 上传软件使用记录
func (a *AppMonitor) uploadUsageLocked(usage []AppUsage, currentTime time.Time) {
	if a.client.offlineMode != 0 {
		logInfo("离线模式，软件使用数据已缓存")
		a.saveBatchToCache(usage)
		return
	}

	if a.client.apiClient == nil {
		logError("API客户端未初始化")
		a.saveBatchToCache(usage)
		return
	}

	logInfo("📤 上报软件使用记录: %d条", len(usage))

	batchSize := 100
	successCount := 0

	for i := 0; i < len(usage); i += batchSize {
		end := i + batchSize
		if end > len(usage) {
			end = len(usage)
		}

		batch := usage[i:end]

		// 转换为map格式用于上报
		serializable := make([]map[string]interface{}, len(batch))
		for j, item := range batch {
			serializable[j] = map[string]interface{}{
				"employee_id":  item.EmployeeID,
				"client_id":    item.ClientID,
				"app_name":     item.AppName,
				"app_path":     item.AppPath,
				"window_title": item.WindowTitle,
				"start_time":   item.StartTime.Format(time.RFC3339),
				"duration":     item.Duration,
				"cpu_avg":      item.CPUAvg,
				"memory_avg":   item.MemoryAvg,
			}
		}

		// 添加重试逻辑
		var err error
		for retry := 0; retry < 3; retry++ {
			if retry > 0 {
				logDebug("重试上报批次 %d-%d (%d/3)", i, end, retry+1)
				time.Sleep(time.Duration(retry) * time.Second)
			}

			_, err = a.client.apiClient.Post("/api/apps/usage", serializable)
			if err == nil {
				break
			}
			logDebug("上报失败: %v", err)
		}

		if err != nil {
			logError("上报软件使用批次失败: %v", err)
			a.saveBatchToCache(batch)
		} else {
			logInfo("✅ 软件使用上报成功: %d条 (批次 %d-%d)", len(batch), i, end)
			successCount += len(batch)
		}
	}

	if successCount > 0 {
		a.lastReportTime = currentTime
	}
}

// saveBatchToCache 保存批次到缓存
func (a *AppMonitor) saveBatchToCache(batch []AppUsage) {
	cacheDir := "cache"
	os.MkdirAll(cacheDir, 0755)

	for _, item := range batch {
		cacheFile := filepath.Join(cacheDir, fmt.Sprintf("app_%d_%d.json",
			item.StartTime.Unix(), time.Now().UnixNano()))
		jsonData, _ := json.Marshal(item)
		os.WriteFile(cacheFile, jsonData, 0644)
	}
}

// ProcessInfo 进程信息
type ProcessInfo struct {
	PID        int
	Name       string
	Exe        string
	CPUPercent float64
	MemoryMB   float64
}

// ForegroundApp 前台应用信息
type ForegroundApp struct {
	PID         int
	Name        string
	Exe         string
	WindowTitle string
	CPUPercent  float64
	MemoryMB    float64
}